Ageller/target gen

alan-geller opened the pull request:
See issue #196.

This PR includes a new rewrite step that generates an abstract base class for a new simulator or emulator class from the intrinsics defined in a Q# file or project.

bettinaheim commented on src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs:
```diff
@@ -0,0 +1,398 @@
+            ``constructor`` name ``(`` [ ("m", ``type`` targetClass) ] ``)`` ``:`` ["m"]
+                    [``public``] 
+                    ``(`` [ (``ident`` "simulator" <-- ``ident`` "m") |> statement ] ``)``
+        let buildAllocClass name =
```
nit: I'd split the larger inner helper functions out into separate private functions so that it is easier to read through the actual buildEpilogue function.

bettinaheim commented on src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs:
```diff
@@ -0,0 +1,398 @@
+        let inType  = SimulationCode.roslynTypeName context sp.Signature.ArgumentType
+        let outType = SimulationCode.roslynTypeName context sp.Signature.ReturnType
+        let propertyType = "Func<" + inType + ", " + outType + ">"
+        let propertyName =
```
Since these need to match the generated C# code, it would be good to split out the corresponding functionality in the C# generation and invoke that one instead.

alan-geller replied:
Done.

bettinaheim commented on src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs:
```diff
@@ -0,0 +1,398 @@
+        match sp.Implementation with
+        | SpecializationImplementation.Intrinsic -> 
+            let returnType  = sp.Signature.ReturnType
+            let inData = ``ident`` "__in__"
```
similarly here.

alan-geller replied:
Dome.

bettinaheim commented on src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs:
```diff
@@ -0,0 +1,398 @@
+                         | QsControlled | QsControlledAdjoint -> 
+                            let controls = 
+                                { 
+                                    VariableName = NonNullable<string>.New "__ctrls__" |> ValidName
```
For this logic you could actually use the SyntaxGenerator.ImmutableQubitArrayWithName. 
nit: I believe the name needs to match the one in fullArgsForSpecialization? If so, I suggest to have the used string in one place and use that in both functions.

alan-geller replied:
The SyntaxGenerator function creates a TYpedExpression, where here I need a LocalVariable, so it doesn't quite work.
The two strings don't need to match, one is used defining the method and the other is used calling the method. I've eliminated other magic strings where they need to match, but since these don't need to match, I don't want the code to force them to match.

bettinaheim commented on src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs:
```diff
@@ -0,0 +1,398 @@
+                                    Range = (QsPositionInfo.Zero, QsPositionInfo.Zero)
+                                }
+                                |> QsTupleItem
+                            match args with
```
I am not following this: Why is this a flattened argument list? 
And as an edge case: for an operation Op taking unit as argument, Controlled Op would actually require a tuple of the control qubits and a unit value.

alan-geller replied:
The flattening  is copied from the C# code gen. 
Controlled Op(), when Op takes Unit, is a bug; I'll fix it and add a unit test. It's a little odd -- a quantum-controlled random number generator, perhaps, but that's weird -- but it should still be processed correctly.

alan-geller replied:
OK, Controlled Op when Op takes no parameters is now working, I think.

bettinaheim replied:
Thanks!

bettinaheim commented on src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs:
```diff
@@ -0,0 +1,398 @@
+        | Generated SelfInverse ->
+            let adjointedBodyName =
+                match sp.Kind with
+                | QsAdjoint           -> "Body"
```
Same here, it would be good to get these from the C# gen.

alan-geller replied:
Done.

bettinaheim commented on src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs:
```diff
@@ -0,0 +1,398 @@
+        let constructors = [ (SimulationCode.buildConstructor context name) ]
+        let nameProperties = [ SimulationCode.buildName name; 
+                               SimulationCode.buildFullName op.FullName ]
+        let baseOp =
```
This function could also be split out in the C# generation and used from there. And actually, I think so can a lot of things here.

alan-geller replied:
Done.

alan-geller commented:
I refactored a bunch of code between this and the C# code gen to reduce the amount of duplication. There's still a little, but much less. I've also replaced most of the "magic strings" with let-bound variables.
I've resolved the comments that I'm confident are addressed.

alan-geller commented on src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs:
```diff
@@ -0,0 +1,398 @@
+            "Microsoft.Quantum.Simulation.Core"
+        ]
+
+    let private intType = ``ident`` "long"
```
I did it slightly differently, but there's now one piece of code for both.

alan-geller commented on src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs:
```diff
@@ -0,0 +1,398 @@
+    let private autoNamespaces =
+        [
+            "System"
+            "Microsoft.Quantum.Core"
```
I just pulled this from the C# code gen. Andres, below, suggests instead using the fully-qualified names and not opening these; I think that's pretty straightforward to do, but I haven't done it yet. I know Intrinsics is referenced at least once.

bettinaheim replied:
Is the intrinsics namespace is referenced for the qubit management?

swernli replied:
I've noticed this in the generated C# as well. The Core and Intrinsic namespaces are always added, even when not used. This doesn't match up with the Q# behavior, where at least the Intrinsic namespace as to be explicitly opened. I like Andres' idea of switching those to be fully-qualified names when called to avoid having these in the `autoNamespaces` list, and I think we should consider doing the same for the C# generation as well. 

Should I file an issue for that?

alan-geller replied:
@swernli That would be great -- that way we can merge this PR and address that issue later.

alan-geller commented on src/Simulation/Microsoft.Quantum.TargetGeneration/RewriteStep.fs:
```diff
@@ -0,0 +1,94 @@
+        member this.ImplementsPostconditionVerification = false
+        member this.ImplementsTransformation = true
+
+        // The precondition is that the TargetQualifiedClassKey has to be present and has to be
```
Done.

alan-geller commented on src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs:
```diff
@@ -0,0 +1,398 @@
+                                ``(`` [``param`` "qbs" ``of`` qubitArrayType] ``)`` [``public``;``abstract``]
+                                ``{`` [] ``}`` :> MemberDeclarationSyntax
+            seq { implOne; implMany }
+        let dumpMachineMethod = ``method`` "void" "DoDumpMachine" ``<<`` ["T"] ``>>``
```
I think the only possible conflict can occur if someone defines a Q# intrinsic in their target file that is called DumpMachine but is in a different namespace than Microsoft.Quantum.Intrinsics. Indeed, in general if someone has two intrinsics of the same name in different Q# namespaces, what I generate will not work.
I can address this by always name-mangling, which would make the standard case ugly, or by name-mangling only in this case, for the second (and third, etc.) instance of the same intrinsic. I'll add that...

alan-geller replied:
Done, with a unit test.

alan-geller commented on src/Simulation/CsharpGeneration.Tests/Circuits/EmulatorTarget.qs:
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// Target definition file for a CHP simulator, for testing target generation.
```
It's correct, it is indeed the gate set for a CHP simulator.

alan-geller replied:
Now moot, since I've stripped down the test cases and this file is no longer there.

alan-geller commented on src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs:
```diff
@@ -0,0 +1,398 @@
+    let private autoNamespaces =
+        [
+            "System"
+            "Microsoft.Quantum.Core"
```
Right now I share this code with the C# code gen. If I'm feeling ambitious I'll go through both files and add the fqns...

alan-geller commented:
Almost forgot; here are the answer's to @anpaz-msft's questions:

1. **What are the end-to-end steps needed to create a new Simulator.**  First create a Q# project that lists all of the intrinsic operations and functions to be provided by the simulator. Compile the project with this extension, which will create a C# file defining an abstract base class. Copy that file into a new C# project, and add another file that implements a subclass of the abstract base that implements all of the abstract "Do*" methods in the base. Compile that C# project into the new simulator. I suspect this can actually be done so that the output of running this extension is a full C# project with all of the above, but I haven't looked into that.
2. **How do you select what Intrinsics goes into the Simulator.**  Any function or operation with a specialization specified as intrinsic is included.
3. **How many msbuild projects are needed? One for the Intrinsics? One for the Simulator base? One for the actual implementation?** I think just two, one Q# project that is used to create the abstract base class file, and one for the base class and the simulator implementation. The Q# project only gets built when you change the set of operations supported by the simulator, so probably not much after the first couple of times.
4. **How do you use the resulting Simulator?**  Just like any other normal simulator. It derives from `SimulatorBase`, and all of the intrinsic operations are internal classes and so should be picked up by the standard `SimulatorBase` mechanism. There shouldn't be any non-standard plumbing involved.

One other note: you can specify an alternate base class, rather than `SimulatorBase`. This should make it easy to define simulators that are based on an existing simulator, like the full-state simulator, but that provide an intrinsic (i.e., emulated) implementation of some operations: you just create your Q# file defining the operations you want to emulate as intrinsic, and follow the process.

One limitation on this currently is that if any specialization is intrinsic, I assume they all are, so you couldn't implement the body and adjoint as intrinsics but have the controlled specialization be implemented in Q#, say. I'm not sure if this is a real problem, but it's something that could be added (mixed Q# and intrinsic implementations) if need be.

alan-geller commented on src/Simulation/CsharpGeneration.Tests/SimulationCodeTests.fs:
```diff
@@ -3813,5 +3814,939 @@ namespace Microsoft.Quantum.Tests.UnitTests
+
+    [<Fact>]
+    let ``target test: CHP-ish`` () =
+        """//------------------------------------------------------------------------------
```
Done. I've also simplified the unit test a lot.

alan-geller commented on src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs:
```diff
@@ -0,0 +1,398 @@
+        // We add a "pragma warning disable 1591" since we don't generate doc comments in our C# code.
+        |> ``pragmaDisableWarning`` 1591
+        |> ``pragmaDisableWarning`` 0162 // unreachable code
+        |> ``pragmaDisableWarning`` 0436 // shadowing existing classes from references
```
I don't know; again, this is pulled from the C# code gen.

bettinaheim replied:
Hm, for the C# generation this is there because we re-generate the relevant library code after monomorphization, which will then shadow the code in the referenced library. I don't think this should be applicable in this case, unless I forget something.

alan-geller commented on src/Simulation/CsharpGeneration.Tests/SimulationCodeTests.fs:
```diff
@@ -3813,5 +3814,939 @@ namespace Microsoft.Quantum.Tests.UnitTests
+    [<Fact>]
+    let ``target test: CHP-ish`` () =
+        """//------------------------------------------------------------------------------
+// <auto-generated>                                                             
```
I stripped the test files down a lot and split them to cover specific cases. I can't come up with a better way to test than text comparison, unfortunately; I've at least simplified it by splitting the expected output into separate text files.

alan-geller commented on src/Simulation/CsharpGeneration.Tests/Tests.CsharpGeneration.fsproj:
```diff
@@ -53,6 +62,7 @@
   <ItemGroup>
     <ProjectReference Include="..\Core\Microsoft.Quantum.Runtime.Core.csproj" />
     <ProjectReference Include="..\CsharpGeneration\Microsoft.Quantum.CsharpGeneration.fsproj" />
+    <ProjectReference Include="..\Microsoft.Quantum.TargetGeneration\Microsoft.Quantum.TargetGeneration.fsproj" />
```
Fixed.

bettinaheim commented:
> Almost forgot; here are the answer's to @anpaz-msft's questions:
> 
> 1. **What are the end-to-end steps needed to create a new Simulator.**  First create a Q# project that lists all of the intrinsic operations and functions to be provided by the simulator. Compile the project with this extension, which will create a C# file defining an abstract base class. Copy that file into a new C# project, and add another file that implements a subclass of the abstract base that implements all of the abstract "Do*" methods in the base. Compile that C# project into the new simulator. I suspect this can actually be done so that the output of running this extension is a full C# project with all of the above, but I haven't looked into that.
> 2. **How do you select what Intrinsics goes into the Simulator.**  Any function or operation with a specialization specified as intrinsic is included.
> 3. **How many msbuild projects are needed? One for the Intrinsics? One for the Simulator base? One for the actual implementation?** I think just two, one Q# project that is used to create the abstract base class file, and one for the base class and the simulator implementation. The Q# project only gets built when you change the set of operations supported by the simulator, so probably not much after the first couple of times.
> 4. **How do you use the resulting Simulator?**  Just like any other normal simulator. It derives from `SimulatorBase`, and all of the intrinsic operations are internal classes and so should be picked up by the standard `SimulatorBase` mechanism. There shouldn't be any non-standard plumbing involved.
> 
> One other note: you can specify an alternate base class, rather than `SimulatorBase`. This should make it easy to define simulators that are based on an existing simulator, like the full-state simulator, but that provide an intrinsic (i.e., emulated) implementation of some operations: you just create your Q# file defining the operations you want to emulate as intrinsic, and follow the process.
> 
> One limitation on this currently is that if any specialization is intrinsic, I assume they all are, so you couldn't implement the body and adjoint as intrinsics but have the controlled specialization be implemented in Q#, say. I'm not sure if this is a real problem, but it's something that could be added (mixed Q# and intrinsic implementations) if need be.

@anpaz-msft @alan-geller 
I think it should be fine to have the core package be a single package containing both the Q# library and the generated abstract base class - I definitively don't want to require any manual "copying over". The only thing that will require a bit of handling in the Sdk is to explicitly exclude the Q# core library dll from the list of Q# references. Then the idea would be that any simulator implementation just references the core package. 
On the mentioned limitation in the last paragraph, that is currently prevented on the Q# side.

anpaz-msft commented:
@alan-geller, @bettinaheim,

Thanks for the answers, a couple of more questions:
- Will the plan be to create a different package with a specific set of Intrinsics for different architectures?
- How do we allow that the same logical intrinsic (say X) defined along multiple packages/architectures?
- At what point does the user needs to select what architecture it is targeting? How is the architecture selected?
- What would happen if someone tries to run a program targeted for one architecture using a simulator for a different architecture?
Also, I might be wrong but I think some extra plumbing might be needed for standalone Q# executables. That can come later.

swernli commented:
@anpaz-msft For the first two points, well-defined and flexible decomposition packages is part of what I'm tackling under #249. I discussed my current plan with Alan to make sure it lines up well with how we want this generation code to work, as well as how we can make it easy for developers to add/remove gates from their custom packages. I think I have a good plan for how to do that in a way that allows the maximum code reuse while still allowing flexibility, such that someone creating a new decomposition package would only have to write the intrinsics/decompositions that are different and can just link to the reset of the individual gate operations they'd like to reuse.

I just picked this work back up and will be continuing out of the feature/decomp branch in the coming days. I'd be happy to get into more details and bounce ideas back and forth regarding this, perhaps in the comment thread on the issue linked above!

alan-geller commented:
@anpaz-msft 
For the last two points:

- At what point does the user needs to select what architecture it is targeting? How is the architecture selected?
- What would happen if someone tries to run a program targeted for one architecture using a simulator for a different architecture?

The answers are tied together. You can run a fully-built Q# program against any simulator that defines the gates that the program winds up calling, so if I took something that I wrote to run against the Toffoli simulator and run it against the full-state simulator, it should work (ignoring speed and memory usage :-)). What I've built here shouldn't change this from working just like it does now.

Still, I think it will be good to select a target as part of building a Q# executable, and have the target definition include not just the list of Q# intrinsics, but also Q# code that implements some "primitives" in terms of those intrinsics, like perhaps implementing T as a Z rotation, that sort of thing.

alan-geller commented:
I just realized, during a conversation with Stefan, that I'm not creating the operation inner classes as subclasses of the proper abstract base classes. This has to be fixed before merging.

alan-geller commented:
> I just realized, during a conversation with Stefan, that I'm not creating the operation inner classes as subclasses of the proper abstract base classes. This has to be fixed before merging.

Fixed.

bettinaheim commented:
@swernli, I think these changes are incorporated and adapted for the new simulator design as part of Scott's work, right? I believe this PR is obsolete, so I'll close it and we can discuss after the holidays what (if anything) of this PR has not yet been incorporated.
